librad/src/uri.rs

```diff
@@ -0,0 +1,535 @@
+    /// of a URN
+    pub fn nss(&self) -> &str {
+        match self {
+            Self::Git => "git",
+            //Self::Pijul => "pijul",
+        }
+    }
+
+    pub fn from_nss(s: &str) -> Option<Self> {
+        match s {
+            "git" => Some(Self::Git),
+            //"pijul" => Some(Self::Pijul),
+            _ => None,
+        }
+    }
+}
+
+pub mod path {
+    use super::*;
+    #[derive(Debug, Error, PartialEq)]
+    #[error("Malformed path: {reasons:?}")]
+    pub struct ParseError {
+        pub reasons: Vec<&'static ViolatesRefFormat>,
+    }
+
+    #[derive(Debug, Error, PartialEq)]
+    pub enum ViolatesRefFormat {
+        #[error("Ends with `.lock`")]
+        EndsWithDotLock,
+
+        #[error("Starts with a dot (`.`)")]
+        StartsWithDot,
+
+        #[error("Contains consecutive dots (`..`)")]
+        ConsecutiveDots,
+
+        #[error("Contains control characters")]
+        ControlCharacters,
+
+        #[error("Contains reserved characters (`~`, `^`, `:`, `?`, `*`, `[`, `\\`)")]
+        ReservedCharacters,
+
+        #[error("Contains `@{{`")] // nb. double-brace is to escape format string
+        AtOpenBrace,
+
+        #[error("Contains consecutive slashes (`//`)")]
+        ConsecutiveSlashes,
+
+        #[error("Consists of only the `@` character")]
+        OnlyAt,
+    }
+}
+
+/// The path component of a [`RadUrn`]
+///
+/// A [`Path`] is also a valid git branch name (as specified in
+/// `git-check-ref-format(1)`).
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Path(String);
+
+impl Path {
+    /// Invalid characters and -sequences acc. to `git-check-ref-format(1)`
+    const REF_FORMAT_RULES: [(&'static str, &'static path::ViolatesRefFormat); 8] = [
+        (r"\.lock$", &path::ViolatesRefFormat::EndsWithDotLock),
+        (r"^\.", &path::ViolatesRefFormat::StartsWithDot),
+        (r"\.\.", &path::ViolatesRefFormat::ConsecutiveDots),
+        (r"[[:cntrl:]]", &path::ViolatesRefFormat::ControlCharacters),
+        (r"[~^:?*\[\\]", &path::ViolatesRefFormat::ReservedCharacters),
+        (r"@[{]", &path::ViolatesRefFormat::AtOpenBrace),
+        (r"//", &path::ViolatesRefFormat::ConsecutiveSlashes),
+        (r"^@$", &path::ViolatesRefFormat::OnlyAt),
+    ];
+
+    pub fn new() -> Self {
+        Self(String::new())
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(String::with_capacity(capacity))
+    }
+
+    pub fn parse<S: AsRef<str>>(s: S) -> Result<Self, path::ParseError> {
+        Self::parse_str(s).map(Path)
+    }
+
+    pub fn append(mut self, other: &Self) -> Self {
+        if !self.0.is_empty() {
+            self.0.push('/')
+        }
+        self.0.push_str(&other.0);
+        self
+    }
+
+    pub fn join<S: AsRef<str>>(self, segment: S) -> Result<Self, path::ParseError> {
+        let segment = Self::parse_str(segment)?;
+        Ok(self.append(&Self(segment)))
```

Comment:
I expected it to join with a separator like `/`. I think this is the usual behaviour of functions called `join` on path-like structures. Is there a reason we don't? Should the functionality be called out in a doc string?

Reply:
It does that, in `append`. I don't actually need `append` anymore, though, so might as well remove it.

Reply:
Ah, `append` is the confusing part then :joy:

>  I don't actually need append anymore, though, so might as well remove it.

Okay :)